### ISSUE TITLE
[Classfication Model] Please add ConvNext Model family

### ISSUE THREAD

**Issue by horlar1** [mlverse/torchvision#210]
ConvNext models are based on this https://arxiv.org/abs/2201.03545 paper.

Please add

- [ ] Convnext tiny
- [ ] Convnext small
- [ ] convnext base
- [ ] convnext large

With their pretrained weights.

The below code works locally for me. however, it will be great to have the official version via the torchvision library.

library(torch)

#' ConvNext Implementation in R using torch
#' This code implements the ConvNext architecture using the torch package in R.
LayerNorm <- nn_module(
  "LayerNorm",
  initialize = function(normalized_shape, eps = 1e-6, data_format = "channels_last") {
    self$weight <- nn_parameter(torch_ones(normalized_shape))
    self$bias <- nn_parameter(torch_zeros(normalized_shape))
    self$eps <- eps
    self$data_format <- data_format
    self$normalized_shape <- normalized_shape
  },
  forward = function(x) {
    if (self$data_format == "channels_last") {
      nnf_layer_norm(x, self$normalized_shape, self$weight, self$bias, self$eps)
    } else if (self$data_format == "channels_first") {
      u <- x$mean(dim = 2, keepdim = TRUE)
      s <- ((x - u)$pow(2))$mean(dim = 2, keepdim = TRUE)
      x <- (x - u) / (s + self$eps)$sqrt()
      x <- self$weight$unsqueeze(2)$unsqueeze(3) * x + self$bias$unsqueeze(2)$unsqueeze(3)
      x
    } else {
      stop("Unsupported data format")
    }
  }
)

Block <- nn_module(
  "Block",
  initialize = function(dim, drop_path = 0, layer_scale_init_value = 1e-6) {
    self$dwconv <- nn_conv2d(dim, dim, kernel_size = 7, padding = 3, groups = dim)
    self$norm <- LayerNorm(dim, eps = 1e-6)
    self$pwconv1 <- nn_linear(dim, 4 * dim)
    self$act <- nn_gelu()
    self$pwconv2 <- nn_linear(4 * dim, dim)
    self$gamma <- if (layer_scale_init_value > 0) {
      nn_parameter(layer_scale_init_value * torch_ones(dim))
    } else {
      NULL
    }
    self$drop_path <- nn_identity()  
  },
  forward = function(x) {
    input <- x
    x <- self$dwconv(x)
    x <- x$permute(c(1, 3, 4, 2))  
    x <- self$norm(x)
    x <- self$pwconv1(x)
    x <- self$act(x)
    x <- self$pwconv2(x)
    if (!is.null(self$gamma)) {
      x <- self$gamma * x
    }
    x <- x$permute(c(1, 4, 2, 3))  
    x <- input + self$drop_path(x)
    x
  }
)

ConvNeXt <- nn_module(
  "ConvNeXt",
  initialize = function(in_chans = 3, num_classes = 1000,
                        depths = c(3, 3, 9, 3), dims = c(96, 192, 384, 768),
                        drop_path_rate = 0., layer_scale_init_value = 1e-6,
                        head_init_scale = 1.) {
    
    self$downsample_layers <- nn_module_list()
    
    # Stem
    stem <- nn_sequential(
      nn_conv2d(in_chans, dims[1], kernel_size = 4, stride = 4),
      LayerNorm(dims[1], eps = 1e-6, data_format = "channels_first")
    )
    self$downsample_layers$append(stem)
    
    for (i in 1:3) {
      self$downsample_layers$append(nn_sequential(
        LayerNorm(dims[i], eps = 1e-6, data_format = "channels_first"),
        nn_conv2d(dims[i], dims[i+1], kernel_size = 2, stride = 2)
      ))
    }
    
    self$stages <- nn_module_list()
    dp_rates <- as.numeric(torch_linspace(0, drop_path_rate, sum(depths)))
    cur <- 1
    for (i in 1:4) {
      blocks <- lapply(1:depths[i], function(j) {
        Block(dims[i], drop_path = dp_rates[cur], layer_scale_init_value = layer_scale_init_value)
      })
      self$stages$append(nn_sequential(!!!blocks))
      cur <- cur + depths[i]
    }
    
    self$norm <- nn_layer_norm(dims[4], eps = 1e-6)
    self$head <- nn_linear(dims[4], num_classes)
  },
  
  forward_features = function(x) {
    for (i in 1:4) {
      x <- self$downsample_layers[[i]](x)
      x <- self$stages[[i]](x)
    }
    x$mean(dim = c(3, 4)) %>% self$norm()
  },
  
  forward = function(x) {
    x <- self$forward_features(x)
    self$head(x)
  }
)

convnext_model_urls <- c(
  "convnext_tiny_224_1k" = paste0(getwd(),"/Python_Models/convnext_tiny_22k_1k_224_weights_only.pth"),
  "convnext_tiny_224_22k" = paste0(getwd(),"/Python_Models/convnext_tiny_22k_weights_only.pth"),
  "convnextsmall_224_22k" = paste0(getwd(),"/Python_Models/convnext_small_22k_224_weights_only.pth"),
  'convnextsmall_224_22k_1k'= paste0(getwd(),"/Python_Models/convnext_small_22k_1k_224_weights_only.pth")
)

.convnext <- function(arch, channels, depths, dims, num_classes, pretrained, progress, ...) {
  if (!is.character(arch) || length(arch) != 1) {
    stop("arch must be a single character string.")
  }

  depths <- as.integer(depths)
  dims <- as.integer(dims)
  channels <- as.integer(channels)
  num_classes <- as.integer(num_classes)

  if (length(depths) != 4 || length(dims) != 4) {
    stop("depths and dims must be vectors of length 4.")
  }

  model <- ConvNeXt(in_chans = channels, depths = depths, dims = dims, num_classes = num_classes)

  if (pretrained) {
    if (!arch %in% names(convnext_model_urls)) {
      stop(paste("Pretrained model for", arch, "is not available."))
    }

    state_dict <- torch::load_state_dict(convnext_model_urls[arch])

    # Interpolate stem weights if input channels differ.sample use cases - satellite images
    conv1_weight <- state_dict[["downsample_layers.0.0.weight"]]

    if (dim(conv1_weight)[2] != channels) {
      old_in_channels <- dim(conv1_weight)[2]
      mean_weight <- conv1_weight$mean(dim = 2, keepdim = TRUE)  

      # Repeat manually using torch_cat
      new_weight_list <- rep(list(mean_weight), channels)
      new_weight <- torch_cat(new_weight_list, dim = 2)  

      new_weight <- new_weight * (old_in_channels / channels)
      state_dict[["downsample_layers.0.0.weight"]] <- new_weight
    }

    model$load_state_dict(state_dict, strict = FALSE)
  }

  model
}

model_convnextTiny_224_22k_1k <- function(pretrained = FALSE, progress = TRUE,channels = 3, ...) {
  .convnext(arch = "convnext_tiny_224_1k" ,channels =channels, depths = c(3, 3, 9, 3), 
            dims = c(96, 192, 384, 768), 
            num_classes=1000, pretrained, progress, ...)
}

model_convnextTiny_224_22k <- function(pretrained = FALSE, progress = TRUE,channels = 3, ...) {
  .convnext("convnext_tiny_224_22k",channels =channels, depths = c(3, 3, 9, 3), 
            dims = c(96, 192, 384, 768), 
            num_classes=21841, pretrained, progress, ...)
}

model_convnextSmall_224_22k_1k <- function(pretrained = FALSE, progress = TRUE,channels = 3, ...) {
  .convnext(arch = "convnextsmall_224_22k_1k" ,channels =channels, depths = c(3, 3, 27, 3), 
            dims = c(96, 192, 384, 768), 
            num_classes=1000, pretrained, progress, ...)
}

model_convnextSmall_224_22k <- function(pretrained = FALSE, progress = TRUE,channels = 3, ...) {
  .convnext("convnextsmall_224_22k",channels =channels, depths = c(3, 3, 27, 3), 
            dims = c(96, 192, 384, 768), 
            num_classes=21841, pretrained, progress, ...)
}
 

model <- model_convnextTiny_224_22k(pretrained = T)
print(model)
An `nn_module` containing 44,615,857 parameters.

── Modules ─────────────────────────────────
• downsample_layers: <nn_module_list> #1,555,872 parameters
• stages: <nn_module_list> #26,262,720 parameters
• norm: <nn_layer_norm> #1,536 parameters
• head: <nn_linear> #16,795,729 parameters
> 
> input <- torch_randn(1, 3, 224, 224)  
> output <- model(input)
> 
> print(output$shape) 
[1]     1 21841
